shared/domain/src/commonMain/kotlin/network/bisq/mobile/client/websocket/WebSocketClient.kt

```diff
@@ -0,0 +1,183 @@
+package network.bisq.mobile.client.websocket
+
+import io.ktor.client.HttpClient
+import io.ktor.client.plugins.websocket.DefaultClientWebSocketSession
+import io.ktor.client.plugins.websocket.webSocketSession
+import io.ktor.client.request.url
+import io.ktor.util.collections.ConcurrentMap
+import io.ktor.websocket.Frame
+import io.ktor.websocket.close
+import io.ktor.websocket.readText
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonTransformingSerializer
+import kotlinx.serialization.json.jsonObject
+import network.bisq.mobile.client.websocket.messages.SubscriptionRequest
+import network.bisq.mobile.client.websocket.messages.SubscriptionResponse
+import network.bisq.mobile.client.websocket.messages.WebSocketEvent
+import network.bisq.mobile.client.websocket.messages.WebSocketMessage
+import network.bisq.mobile.client.websocket.messages.WebSocketRequest
+import network.bisq.mobile.client.websocket.messages.WebSocketResponse
+import network.bisq.mobile.client.websocket.subscription.ModificationType
+import network.bisq.mobile.client.websocket.subscription.Topic
+import network.bisq.mobile.client.websocket.subscription.WebSocketEventObserver
+import network.bisq.mobile.domain.data.BackgroundDispatcher
+import network.bisq.mobile.utils.Logging
+import network.bisq.mobile.utils.createUuid
+
+class WebSocketClient(
```

Comment:
This is overall a great implementation of a WebSocketClient, quite robust and thread safe.

Just a few optimisations I would suggest to add in:

 1. handle exceptions for `startListening` and `send` calls
 2. Same as in the Handler class, code is needed for handling errors on `request` calls
 3. provide a reconnect method and some way to try automatic reconnect if connection gets lost (This one we can create an issue and do later)
 4. implement unsubscribe to avoid mem leaks (same as number 3)